service/service.go

```diff
@@ -42,52 +67,103 @@ func NewService(cfg *Config) (*Service, error) {
 			case <-s.roundsTicker():
 			}
 
-			if len(s.activeRound.commitments) == 0 && !s.cfg.ExecuteEmpty {
+			if len(s.openRound.commitments) == 0 && !s.cfg.ExecuteEmpty {
 				continue
 			}
 
-			s.prevRound = s.activeRound
+			s.prevRound = s.openRound
 
 			roundId++
-			s.activeRound = s.newRound(roundId)
+			s.openRound = s.newRound(roundId)
 			log.Infof("round %v opened", roundId)
 
 			// Close previous round and execute it.
 			go func() {
+				// TODO(moshababo): apply safe concurrency
 				r := s.prevRound
+				s.executingRounds[r.id] = r
+
 				err := r.close()
 				if err != nil {
 					log.Error(err)
 				}
 				log.Infof("round %v closed, executing...", r.id)
-				phi, err := r.execute()
+				err = r.execute()
 				if err != nil {
 					log.Error(err)
 				}
-				log.Infof("round %v executed, phi=%v", r.id, phi)
+
+				delete(s.executingRounds, r.id)
+				s.executedRounds[r.id] = r
+				log.Infof("round %v executed, phi=%v", r.id, r.nip.Phi)
 			}()
 		}
 	}()
 
 	return s, nil
 }
 
-func (s *Service) SubmitCommitment(c []byte) (*submitResponse, error) {
-	r := s.activeRound
+func (s *Service) Info() *InfoResponse {
+	res := new(InfoResponse)
+	res.openRoundId = s.openRound.id
+
+	ids := make([]int, 0, len(s.executingRounds))
+	for id := range s.executingRounds {
+		ids = append(ids, id)
+	}
+	res.executingRoundsId = ids
+
+	ids = make([]int, 0, len(s.executedRounds))
+	for id := range s.executedRounds {
+		ids = append(ids, id)
+	}
+	res.executedRoundsId = ids
+
+	return res
+}
+
+func (s *Service) RoundInfo(roundId int) (*RoundInfoResponse, error) {
+	r := s.rounds[roundId]
+	if r == nil {
+		return nil, ErrRoundNotFound
+	}
+
+	res := new(RoundInfoResponse)
+	res.opened = r.opened
+	res.executeStart = r.executeStart
+	res.executeEnd = r.executeEnd
+	res.numOfCommitments = len(r.commitments)
+	res.merkleRoot = r.merkleRoot
+	res.nip = r.nip
+
+	return res, nil
+}
+
+func (s *Service) round(roundId int) (*round, error) {
+	r := s.rounds[roundId]
+	if r == nil {
+		return nil, ErrRoundNotFound
+	}
+
+	return r, nil
+}
+
+func (s *Service) SubmitCommitment(c []byte) (*SubmitCommitmentResponse, error) {
+	r := s.openRound
 	err := r.submitCommitment(c)
 	if err != nil {
 		return nil, err
 	}
 
-	res := new(submitResponse)
+	res := new(SubmitCommitmentResponse)
```

Comment:
no actual need for new, you can also refactor the two lines to ine using `res := SubmitCommitmentResponse {roundId : r.id}`

Reply:
Do you mean `&SubmitCommitmentResponse {roundId : r.id}`?
If so, since they are equivalent, you mean that that’s the preferred style?
I'm asking because I used `new` in all other pointer constructs as well.

Reply:
Yes, using `new` keyword is not common in go apps, but it is not important to refactor it. Just wanted to mention the alternative that is also more compact